Userland/Libraries/LibSoftGPU/Clipper.cpp

```diff
@@ -97,49 +99,76 @@ void Clipper::clip_points_against_frustum(Vector<GPU::Vertex>& vertices)
+
     auto from_within_plane = point_within_clip_plane<plane>(from.clip_coordinates);
     auto to_within_plane = point_within_clip_plane<plane>(to.clip_coordinates);
     if (!from_within_plane && !to_within_plane)
         return false;
     if (!from_within_plane)
-        from = clip_intersection_point<plane>(from, to);
+        from = clip_intersection_point<plane>(from, to, clip_plane_eqn);
     else if (!to_within_plane)
-        to = clip_intersection_point<plane>(from, to);
+        to = clip_intersection_point<plane>(from, to, clip_plane_eqn);
     return true;
 }
 
 bool Clipper::clip_line_against_frustum(GPU::Vertex& from, GPU::Vertex& to)
 {
-    return constrain_line_within_plane<ClipPlane::LEFT>(from, to)
-        && constrain_line_within_plane<ClipPlane::RIGHT>(from, to)
-        && constrain_line_within_plane<ClipPlane::TOP>(from, to)
-        && constrain_line_within_plane<ClipPlane::BOTTOM>(from, to)
-        && constrain_line_within_plane<ClipPlane::NEAR>(from, to)
-        && constrain_line_within_plane<ClipPlane::FAR>(from, to);
+    return constrain_line_within_plane<ClipPlane::Left>(from, to)
+        && constrain_line_within_plane<ClipPlane::Right>(from, to)
+        && constrain_line_within_plane<ClipPlane::Top>(from, to)
+        && constrain_line_within_plane<ClipPlane::Bottom>(from, to)
+        && constrain_line_within_plane<ClipPlane::Near>(from, to)
+        && constrain_line_within_plane<ClipPlane::Far>(from, to);
 }
 
 void Clipper::clip_triangle_against_frustum(Vector<GPU::Vertex>& input_verts)
 {
     // FIXME C++23. Static reflection will provide looping over all enum values.
-    clip_plane<ClipPlane::LEFT>(input_verts, m_vertex_buffer);
-    clip_plane<ClipPlane::RIGHT>(m_vertex_buffer, input_verts);
-    clip_plane<ClipPlane::TOP>(input_verts, m_vertex_buffer);
-    clip_plane<ClipPlane::BOTTOM>(m_vertex_buffer, input_verts);
-    clip_plane<ClipPlane::NEAR>(input_verts, m_vertex_buffer);
-    clip_plane<ClipPlane::FAR>(m_vertex_buffer, input_verts);
+    clip_plane<ClipPlane::Left>(input_verts, m_vertex_buffer, clip_plane_eqns[0]);
+    clip_plane<ClipPlane::Right>(m_vertex_buffer, input_verts, clip_plane_eqns[1]);
+    clip_plane<ClipPlane::Top>(input_verts, m_vertex_buffer, clip_plane_eqns[2]);
+    clip_plane<ClipPlane::Bottom>(m_vertex_buffer, input_verts, clip_plane_eqns[3]);
+    clip_plane<ClipPlane::Near>(input_verts, m_vertex_buffer, clip_plane_eqns[4]);
+    clip_plane<ClipPlane::Far>(m_vertex_buffer, input_verts, clip_plane_eqns[5]);
+}
+
+void Clipper::clip_triangle_against_user_defined(Vector<GPU::Vertex>& input_verts, Vector<FloatVector4>& user_planes)
+{
+    // FIXME: Also implement user plane support for points and lines
+    auto& in = input_verts;
+    auto& out = m_vertex_buffer;
+    for (auto const& plane : user_planes) {
+        clip_plane<ClipPlane::User>(in, out, plane);
+        auto& temp = in;
+        in = out;
+        out = temp;
```

Comment:
Please use `swap()` from `AK/StdLibExtras.h`

Reply:
Ah neat! Changed in latest push